pkg/cmd/factory/factory_impl.go

```diff
@@ -0,0 +1,148 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package factory
+
+import (
+	"crypto/tls"
+	"strings"
+	"time"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/ticdc/cdc/kv"
+	cmdconetxt "github.com/pingcap/ticdc/pkg/cmd/context"
+	"github.com/pingcap/ticdc/pkg/security"
+	"github.com/pingcap/ticdc/pkg/version"
+	pd "github.com/tikv/pd/client"
+	"go.etcd.io/etcd/clientv3"
+	etcdlogutil "go.etcd.io/etcd/pkg/logutil"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/backoff"
+)
+
+type factoryImpl struct {
+	clientGetter ClientGetter
+}
+
+// NewFactory creates a client build factory.
+func NewFactory(clientGetter ClientGetter) Factory {
+	if clientGetter == nil {
+		panic("attempt to instantiate factory with nil clientGetter")
+	}
+	f := &factoryImpl{
+		clientGetter: clientGetter,
+	}
+
+	return f
+}
+
+func (f *factoryImpl) ToTLSConfig() (*tls.Config, error) {
+	return f.clientGetter.ToTLSConfig()
+}
+
+func (f *factoryImpl) ToGRPCDialOption() (grpc.DialOption, error) {
+	return f.clientGetter.ToGRPCDialOption()
+}
+
+func (f *factoryImpl) GetPdAddr() string {
+	return f.clientGetter.GetPdAddr()
+}
+
+func (f *factoryImpl) GetCredential() *security.Credential {
+	return f.clientGetter.GetCredential()
+}
+
+func (f *factoryImpl) EtcdClient() (*kv.CDCEtcdClient, error) {
+	ctx := cmdconetxt.GetDefaultContext()
+
+	tlsConfig, err := f.ToTLSConfig()
+	if err != nil {
+		return nil, err
+	}
+	grpcTLSOption, err := f.ToGRPCDialOption()
+	if err != nil {
+		return nil, err
+	}
+
+	logConfig := etcdlogutil.DefaultZapLoggerConfig
+	logConfig.Level = zap.NewAtomicLevelAt(zapcore.ErrorLevel)
+	pdEndpoints := strings.Split(f.GetPdAddr(), ",")
+
+	etcdClient, err := clientv3.New(clientv3.Config{
+		Context:     ctx,
+		Endpoints:   pdEndpoints,
+		TLS:         tlsConfig,
+		LogConfig:   &logConfig,
+		DialTimeout: 30 * time.Second,
+		DialOptions: []grpc.DialOption{
+			grpcTLSOption,
+			grpc.WithBlock(),
+			grpc.WithConnectParams(grpc.ConnectParams{
+				Backoff: backoff.Config{
+					BaseDelay:  time.Second,
+					Multiplier: 1.1,
+					Jitter:     0.1,
+					MaxDelay:   3 * time.Second,
+				},
+				MinConnectTimeout: 3 * time.Second,
+			}),
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	client := kv.NewCDCEtcdClient(ctx, etcdClient)
+	return &client, nil
+}
+
+func (f factoryImpl) PdClient() (pd.Client, error) {
+	ctx := cmdconetxt.GetDefaultContext()
+
+	credential := f.GetCredential()
+	grpcTLSOption, err := f.ToGRPCDialOption()
+	if err != nil {
+		return nil, err
+	}
+
+	pdAddr := f.GetPdAddr()
+	pdEndpoints := strings.Split(pdAddr, ",")
+
+	pdClient, err := pd.NewClientWithContext(
+		ctx, pdEndpoints, credential.PDSecurityOption(),
+		pd.WithGRPCDialOptions(
```

Comment:
Sorry, I'm a little confused about what you mean. Could you go into more detail?

Reply:
I mean the following, so that we can record PD gRPC calls.

https://github.com/pingcap/ticdc/blob/675aea7ad8045298ac88dad5913845c363c53326/cdc/kv/client.go#L275-L276

Reply:
Got it! I will add it.